subxt/src/events/filter_events.rs

```diff
@@ -0,0 +1,418 @@
+    // impl `Stream<Item = Result<Events<'a, T, Evs>, BasicError>> + Unpin + 'a`.
+    sub: Sub,
+    // Each time we get Events from our subscription, they are stored here
+    // and iterated through in future stream iterations until exhausted.
+    events: Option<
+        Box<
+            dyn Iterator<
+                    Item = Result<
+                        FilteredEventDetails<T::Hash, Filter::ReturnType>,
+                        BasicError,
+                    >,
+                > + 'a,
+        >,
+    >,
+}
+
+impl<'a, Sub: 'a, T: Config, Filter: EventFilter> Unpin
+    for FilterEvents<'a, Sub, T, Filter>
+{
+}
+
+impl<'a, Sub: 'a, T: Config, Filter: EventFilter> FilterEvents<'a, Sub, T, Filter> {
+    pub(crate) fn new(sub: Sub) -> Self {
+        Self { sub, events: None }
+    }
+}
+
+impl<'a, Sub, T, Evs, Filter> Stream for FilterEvents<'a, Sub, T, Filter>
+where
+    Sub: Stream<Item = Result<Events<'a, T, Evs>, BasicError>> + Unpin + 'a,
+    T: Config,
+    Evs: Decode + 'static,
+    Filter: EventFilter,
+{
+    type Item = Result<FilteredEventDetails<T::Hash, Filter::ReturnType>, BasicError>;
+    fn poll_next(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        loop {
+            // Drain the current events we're iterating over first:
+            if let Some(events_iter) = self.events.as_mut() {
+                match events_iter.next() {
+                    Some(res) => return Poll::Ready(Some(res)),
+                    None => {
+                        self.events = None;
+                    }
+                }
+            }
+
+            // Wait for new events to come in:
+            match futures::ready!(self.sub.poll_next_unpin(cx)) {
+                None => return Poll::Ready(None),
+                Some(Err(e)) => return Poll::Ready(Some(Err(e))),
+                Some(Ok(events)) => {
+                    self.events = Some(Filter::filter(events));
+                }
+            };
+        }
+    }
+}
+
+/// This is returned from the [`FilterEvents`] impl of [`Stream`]. It contains
+/// some type representing an event we've filtered on, along with couple of additional
+/// pieces of information about that event.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct FilteredEventDetails<BlockHash, Evs> {
+    /// During which [`Phase`] was the event produced?
+    pub phase: Phase,
+    /// Hash of the block that this event came from.
+    pub block_hash: BlockHash,
+    /// A type containing an event that we've filtered on.
+    /// Depending on the filter type, this may be a tuple
+    /// or a single event.
+    pub event: Evs,
+}
+
+/// This trait is implemented for tuples of Event types; any such tuple (up to size 8) can be
+/// used to filter an event subscription to return only the specified events.
+pub trait EventFilter: private::Sealed {
```

Comment:
Maybe move this type higher up in the file, given it's used extensively by the rest of the code?

Reply:
I had a look, and `FilterEvents`/`FilteredEventDetails` both feel like they deserve to be higher to me, since they are the "api" we expose here, and `EventFilter` is just a part of how it works (albeit an important one!).